### PR TITLE
Jakarta Faces 4.0

### PR DESCRIPTION
**Pull request by arjantijms**
## Specification PR template
When creating a specification project release review, create PRs with the content defined as follows.

Include the following in the PR:
- [X] A directory in the form wombat/x.y where x.y is the release major.minor version, and the directory contains the following.
- [X] Specification PDF in the form of jakarta-wombat-spec-x.y.pdf
- [X] Specification HTML in the form of jakarta-wombat-spec-x.y.html
- [X] A specification page named _index.md following the template at:
      https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md
- [X] For a Progress Review, that sufficient progress has been made on a Compatible Implementation and TCK, to ensure that the spec is implementable and testable.
- [ ] For a [Release Review](https://www.eclipse.org/projects/handbook/#release-review), a summary that a Compatible Implementation is complete, passes the TCK, and that the TCK includes sufficient coverage of the specification. The TCK users guide MUST include the instructions to run the compatible implementations used to validate the release.
Instructions MAY be by reference.
   - [x] Updated release record
   - [x] Generated IP Log
   - [x] Email to PMC
   - [x] Start release review by emailing EMO 
- [X] The URL of the OSSRH staging repository for the api, javadoc:
      https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/faces/jakarta.faces-api/4.0.0/
- [X] The URL of the staging directory on downloads.eclipse.org for the proposed EFTL TCK binary:
      https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee10/staged/eftl/jakarta-faces-tck-4.0.0.zip
- [X] The URL of the compatibility certification request issue:
      https://github.com/jakartaee/faces/issues/1673
- [X] Specification JavaDoc in the wombat/x.y/apidocs directory. 
If desired, an optional second PR can be created to contain just the JavaDoc in the `apidocs` directory.

Note: If any item does not apply, check it and mark N/A below it.

Signed-off-by: arjantijms [arjan.tijms@gmail.com](mailto:arjan.tijms@gmail.com)

### CODE REVIEW

**Comment by kazumura** [comment]
# Spec Review Checklist

1. Spec PR
  - [x] PR uses [template](https://github.com/jakartaee/specifications/blob/master/pull_request_template.md)
  - [ ] Directory of form {spec}/x.y
  - [ ] PDF of form jakarta-{spec}-spec-x.y.pdf ("-spec" preferred but not required)
  - [ ] HTML of form jakarta-{spec}-spec-x.y.html ("-spec" preferred but not required)
  - [ ] Index page {spec}/x.y/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md)
  - [ ] Index page {spec}/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_index_template.md)
  - [ ] No other files (e.g., no jakarta_ee_logo_schooner_color_stacked_default.png)
  - [ ] Staging repository link of the form https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/{spec}/jakarta.{spec}-api/x.y.z/
  - [ ] EFTL TCK link of the form http://download.eclipse.org/.../+.zip
  - [ ] Compatibility certification link of the form https://github.com/eclipse-ee4j/{project}/#{issue}
  - [ ] (Optional) Second PR for just apidocs

2. _index.md
  - [ ] Link to project release plan of the form https://projects.eclipse.org/projects/ee4j.{spec}/releases/x.y/plan
  - [ ] Link to spec pdf
  - [ ] Link to spec html
  - [ ] Link to apidocs
  - [ ] Link to final TCK download zip file of the form https://download.eclipse.org/jakartaee/{spec}/x.y/*{spec}-tck-x.y.z.zip (The folder path is required, the file name pattern is preferred.)
  - [ ] Link to API jar file of the form https://search.maven.org/artifact/jakarta.{spec}/jakarta.{spec}-api/x.y.z/jar
  - [ ] Name of and link to at least one Compatible Implementation
  - [ ] Review summary statement to ensure it has been updated to reflect the work accomplished and does not contain proposal statements as might be found in Plan Review statement.

3. javadocs
  - [ ] Footer contains Eclipse copyright and link to license
  - [ ] ESFL license is included, usually as doc-files/speclicense.html
  - [ ] no META-INF directory in PR
  - [ ] javadocs-jar artifact matches apidocs (optional for this release)

4. Spec PDF
  - [ ] Correct spec title
  - [ ] Version number of the form x.y, not x.y.z
  - [ ] Correct Eclipse copyright line
  - [ ] No DRAFT or SNAPSHOT
  - [ ] Correct Logo

5. Spec HTML
  - [ ] Same as PDF

6. TCK zip file
  - [ ] README file (optional for this release)
  - [ ] EFTL license file, preferably named LICENSE.md
  - [ ] User's Guide (or equivalent documentation)
  - [ ] How to test the Compatible Implementation(s) listed in _index.md above with the TCK (may be in UG)

7. TCK User's Guide (or equivalent documentation)
  - [ ] Software requirements listed
  - [ ] Installation and configuration described
  - [ ] How to run tests
  - [ ] Where to file challenges

8. Compatibility certification request
  - [ ] Request follows [template](https://github.com/jakartaee/specification-committee/blob/master/compatibility-certification-request.md)
  - [ ] SHA-256 fingerprint matches staged TCK zip file
  - [ ] Request issue has `certification` label.

9. TCK results summary
  - [ ] Page is hosted by Compatible Implementation project
  - [ ] Includes all information from certification request
  - [ ] Summary includes number of tests passed, failed, errors
  - [ ] SHA-256 fingerprint matches staged TCK zip file on cert request

10. If a Release Review is required, the specification project team contacts the EMO to initiate the review by sending an email to emo@eclipse.org.
    (A Release Review is not required if the current release is a Service Release based on a previously successful Major or Minor
    release as indicated by a release record on the project's Releases page, e.g., the [Jakarta Servlet releases page](https://projects.eclipse.org/projects/ee4j.servlet/reviews).)
    - [ ] An issue will be created by the EMO to track the release review.
    - [ ] The specification project team requests approval for the release from the PMC by sending an email to ee4j-pmc@eclipse.org.
    - [ ] The specification project team then delivers an IP Log to the IP Team for their review as described in https://www.eclipse.org/projects/handbook/#pmi-commands-iplog.

11. Update Jakarta EE API jar
  - [ ] Update the Jakarta EE API jar by submitting a PR to the jakartaee-api project that updates the version number of your API jar file.

**Comment by arjantijms** [comment]
Build failed because of at least one invalid filename being generated for the jsdocs. We're currently looking at this.

**Comment by BalusC** [comment]
> Build failed because of at least one invalid filename being generated for the jsdocs. We're currently looking at this.

Fixed in https://github.com/eclipse-ee4j/mojarra/pull/5057

**Review by kazumura — COMMENTED** [review]
@arjantijms 
The URL of the OSSRH staging repo must be  
 the form "https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/{spec}/jakarta.{spec}-api/x.y.z/".
No "-M5" is not needed.

**Comment by arjantijms** [comment]
> No "-M5" is not needed.

Thanks, I removed the -M5. It's now:

https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/faces/jakarta.faces-api/4.0.0/

**Comment by kazumura** [comment]
@arjantijms 
What timeline do you expect to be ready for TCK and CCR ?

**Comment by arjantijms** [comment]
> What timeline do you expect to be ready for TCK and CCR ?

That's a good question, the spec and API haven been ready obviously since this PR was done, and we're working on the  (new) TCK and getting the old TCK to run as much as possible again via this issue: https://github.com/jakartaee/faces/issues/1600

We kind of agreed that when the deadline is near, whatever doesn't run by then will get excluded, and we'll attempt to include it again for the next release.

cc/ @BalusC @tandraschko

**Comment by lukasj** [comment]
visiting link for `pt:` at https://deploy-preview-459--jakartaee-specifications.netlify.app/specifications/faces/4.0/vdldoc/pt/tld-frame.html - vdldoc for pt - leads to HTTP 404 (just saying...)

**Comment by arjantijms** [comment]
Indeed. All the other ones work. Have to take a look at that. cc @BalusC

**Comment by arjantijms** [comment]
IP log/CQ: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=24063

**Comment by arjantijms** [comment]
Email to PMC: https://eclipse.org/lists/ee4j-pmc/msg03322.html

**Comment by arjantijms** [comment]
Release review tracking issue: https://gitlab.eclipse.org/eclipsefdn/emo-team/emo/-/issues/278

**Comment by arjantijms** [comment]
@lukasj 
> vdldoc for pt - leads to HTTP 404 (just saying...)

Thanks, really have no idea what it's about. It shouldn't be there, so much is clear. And actually the entire "pt" should not be there. I don't think this is blocking the ballot or release though, just a small error.

**Comment by arjantijms** [comment]
@kazumura can you take a look again? Thanks!

File: faces/4.0/vdldoc/pt/tld-frame.html [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        <h1 class="bar">
+            <a href="tld-summary.html" target="tagFrame">pt</a>
+        </h1>
+        <div class="indexContainer">
+            <h2 title="Tags">Tags</h2>
+            <ul title="Tags">
+                <li>
+                    <a href=" .html" target="tagFrame">pt: </a>
+                </li>
+            </ul>
```
Comment (BalusC) [inline faces/4.0/vdldoc/pt/tld-frame.html:40]: Deze <ul> verwijderen.

**Comment by kazumura** [comment]
@arjantijms 
Thanks for submitting CCR. I have left some comments in [CCR issue](https://github.com/jakartaee/faces/issues/1673).

File: faces/4.0/_index.md [hunk 2]
```diff
@@ -1,56 +1,66 @@
...
+* [Jakarta Faces 4.0 Renderkitdoc](./renderkitdoc)
+* [Jakarta Faces 4.0 VDLDoc](./vdldoc)
+* [Jakarta Faces 4.0 TCK](https://download.eclipse.org/jakartaee/faces/4.0/jakarta-faces-tck-4.0.0.zip)([sig](https://download.eclipse.org/jakartaee/faces/4.0/jakarta-faces-tck-3.0.0.zip.sig),[sha](https://download.eclipse.org/jakartaee/faces/4.0/jakarta-faces-tck-4.0.0.zip.sha256),[pub](https://raw.githubusercontent.com/jakartaee/specification-committee/master/jakartaee-spec-committee.pub))
+* Maven coordinates
+  * [jakarta.faces:jakarta.faces-api:jar:4.0.0](https://search.maven.org/artifact/jakarta.faces/jakarta.faces-api/4.0.0/jar)
+
+
+# Compatible Implementations
+
+* [Mojarra 4.0.0-M5](https://github.com/eclipse-ee4j/mojarra/releases/download/4.0.0-M5-RELEASE/jakarta.faces-4.0.0-M5.jar)
```
Comment (Pandrex247) [inline faces/4.0/_index.md:58]: Should this not align with the version on the CCR?
```suggestion
* [Mojarra 4.0.0-M7](https://github.com/eclipse-ee4j/mojarra/releases/download/4.0.0-M7-RELEASE/jakarta.faces-4.0.0-M7.jar)
```

File: faces/4.0/_index.md [hunk 3]
```diff
@@ -1,56 +1,66 @@
...
 
 * [Jakarta Faces 4.0 Release Record](https://projects.eclipse.org/projects/ee4j.faces/releases/4.0)
+    * [Jakarta EE Platform 10 Release Plan](https://eclipse-ee4j.github.io/jakartaee-platform/jakartaee10/JakartaEE10#jakarta-ee-10-release-plan)
+* [Jakarta Faces 4.0 Specification Document](./jakarta-faces-4.0.pdf) (PDF)
+* [Jakarta Faces 4.0 Specification Document](./jakarta-faces-4.0.html) (HTML)
+* [Jakarta Faces 4.0 Javadoc](./apidocs)
+* [Jakarta Faces 4.0 Jsdoc](./jsdoc)
+* [Jakarta Faces 4.0 Renderkitdoc](./renderkitdoc)
+* [Jakarta Faces 4.0 VDLDoc](./vdldoc)
+* [Jakarta Faces 4.0 TCK](https://download.eclipse.org/jakartaee/faces/4.0/jakarta-faces-tck-4.0.0.zip)([sig](https://download.eclipse.org/jakartaee/faces/4.0/jakarta-faces-tck-3.0.0.zip.sig),[sha](https://download.eclipse.org/jakartaee/faces/4.0/jakarta-faces-tck-4.0.0.zip.sha256),[pub](https://raw.githubusercontent.com/jakartaee/specification-committee/master/jakartaee-spec-committee.pub))
```
Comment (Pandrex247) [inline faces/4.0/_index.md:51]: That's probably a broken link?
```suggestion
* [Jakarta Faces 4.0 TCK](https://download.eclipse.org/jakartaee/faces/4.0/jakarta-faces-tck-4.0.0.zip)([sig](https://download.eclipse.org/jakartaee/faces/4.0/jakarta-faces-tck-4.0.0.zip.sig),[sha](https://download.eclipse.org/jakartaee/faces/4.0/jakarta-faces-tck-4.0.0.zip.sha256),[pub](https://raw.githubusercontent.com/jakartaee/specification-committee/master/jakartaee-spec-committee.pub))
```

File: faces/4.0/_index.md (hunk 2 above)
Comment (arjantijms) [reply faces/4.0/_index.md:58]: Probably best indeed. From memory they both passed, but let's stay consistent. Thanks!

File: faces/4.0/_index.md (hunk 3 above)
Comment (arjantijms) [reply faces/4.0/_index.md:51]: Thanks again! Should indeed be 4 of course.

**Comment by JanWesterkamp-iJUG** [comment]
@arjantijms I did a dependency analysis with jQA of the last released 4.0.0 Faces version in the Staging repo. There are still some open issues open:

- Faces is affected by the CDI&Transactions cyclic dependency issue, which will require new versions of these specs being referenced
- Transactions has a RC1 reference in a proposed final version
- the Mojarra version points to 4.0.0-M7 - this is fixed in Faces 4.0.0-SNAPSHOT only, but not released as an artifact yet
- Maven uses old Java SE (separate) source and traget level configuration - instead the Java SE/Maven release config property should be used

Details could be found here:
[jakarta-ee-dependencies_PlatformDependencyVersionsByArtifact.csv](https://github.com/jakartaee/specifications/files/8699783/jakarta-ee-dependencies_PlatformDependencyVersionsByArtifact.csv)

jakarta-ee-dependencies_PlatformDependencyDiagram
jakarta-ee-dependencies_SpecificationDependencyDiagram

**Comment by arjantijms** [comment]
@JanWesterkamp-iJUG Thanks, I'll stage an updated jar soon.

**Comment by Emily-Jiang** [comment]
@arjantijms please see my comments on this release. A few more things are to be fixed.
1. I noticed the [spec doc](https://deploy-preview-459--jakartaee-specifications.netlify.app/specifications/faces/4.0/jakarta-faces-4.0.html#changes-between-4-0-and-3-0) has missing chapters:
Changes between 4.0 and 3.0

TBD
Backward Compatibility with Previous Versions

TBD
This section needs to be populated before the release.

2. In the spec doc,  the copyright statement is incorrect:
It should be Copyright (c) 2019, 2022 Eclipse Foundation.

instead of
Copyright (c) 2022 Eclipse Foundation.

3. https://deploy-preview-459--jakartaee-specifications.netlify.app/specifications/faces/4.0/renderkitdoc/
has this copyright
Copyright � 2002-2010 Oracle America, Inc. All Rights Reserved.
I guess it is wrong.

4. No copyright for
https://deploy-preview-459--jakartaee-specifications.netlify.app/specifications/faces/4.0/vdldoc/
https://deploy-preview-459--jakartaee-specifications.netlify.app/specifications/faces/4.0/jsdoc/

**Comment by arjantijms** [comment]
Item 3 and 4 have been addressed as well.

File: faces/4.0/renderkitdoc/HTML_BASIC/renderkit-summary.html [hunk 4]
```diff
@@ -454,6 +454,6 @@ <H2>HTML_BASIC RenderKit </H2><BR>
 
 <HR>
 
-Copyright � 2002-2010 Oracle America, Inc. All Rights Reserved.
+Copyright &#169; 2002-2010 Oracle America, Inc. All Rights Reserved.
```
Comment (Emily-Jiang) [inline faces/4.0/renderkitdoc/HTML_BASIC/renderkit-summary.html:457]: Are you sure about this copyright? I thought it needs to be under EF license.

**Comment by Emily-Jiang** [comment]
@arjantijms I believe you also need to have the following license sentence for the 3 javadocs. See [here](https://deploy-preview-459--jakartaee-specifications.netlify.app/specifications/faces/4.0/apidocs/) 
Use is subject to [license terms](https://deploy-preview-459--jakartaee-specifications.netlify.app/specifications/faces/4.0/apidocs/doc-files/speclicense.html).

File: faces/4.0/renderkitdoc/HTML_BASIC/renderkit-summary.html (hunk 4 above)
Comment (arjantijms) [reply faces/4.0/renderkitdoc/HTML_BASIC/renderkit-summary.html:457]: Not sure no, I was wondering about it too as I updated it. It was inherited like that. I'll change it to the regular Eclipse one.

**Comment by arjantijms** [comment]
> I believe you also need to have the following license sentence for the 3 javadocs.

Yes, that's a good idea. See latest update.

**Comment by Emily-Jiang** [comment]
@arjantijms thanks for fixing these quickly. I think just one more to go:
https://deploy-preview-459--jakartaee-specifications.netlify.app/specifications/faces/4.0/renderkitdoc/
You need to use the same EF license and copyright as the others, I think. At the moment, it still says "Oracle xxx".

**Comment by arjantijms** [comment]
> I think. At the moment, it still says "Oracle xxx".

At the bottom of the page I see:

```
Comments to: [faces-dev@eclipse.org](mailto:faces-dev@eclipse.org).
Copyright © 2019, 2022 Eclipse Foundation. All rights reserved.
Use is subject to [license terms]
```

Screenshot 2022-05-17 at 00 37 51

Where did you see Oracle?

**Comment by Emily-Jiang** [comment]
Sorry, I was not too clear. When you clicked any of the links on the left, a new page will open. On the bottom of the page, you will see the wrong footer. e.g. when you click `[jakarta.faces.Button](https://deploy-preview-459--jakartaee-specifications.netlify.app/specifications/faces/4.0/renderkitdoc/HTML_BASIC/jakarta.faces.Commandjakarta.faces.Button.html)`


image

**Comment by arjantijms** [comment]
> Sorry, I was not too clear.

Ah I see, sorry for not noticing this before. The tools to generate all those different types of docs for Faces are a bit of a nightmare to work with. Hopefully it's now finally fully fixed ;)

**Comment by Emily-Jiang** [comment]
Thanks @arjantijms ! Looks good!

**Comment by kazumura** [comment]
The ballot has completed.
https://www.eclipse.org/lists/jakarta.ee-spec/msg02598.html

@arjantijms  Thanks for your quick actions.

**Review by kazumura — APPROVED** [review]

**Pull request merged**